### DISCUSSION TITLE
AxRuntime v41.5 available

### DISCUSSION THREAD

**Discussion by deadwood2** [deadw00d/AROS#83] [Announcements]
New release of AxRuntime is available. It is a small, bugfixing release and preparation for first beta of Intuition Window Manager. More information here: https://github.com/deadw00d/AROS/releases/tag/AxRT_41.5

**Comment by xgnusr** [comment]
Dear Sir:

I have tried Final Writer on Linux using its incredible AXRuntime tool, it seemed wonderful to me.

Now, I am interested in implementing Intuition + Scalos using Axruntime.

Could you provide me with a small tutorial on how to do it?

I await your news.

Very thankful:

Fco. Morales

**Reply by deadwood2** [reply]
Hello,

Thanks for testing AxRuntime. With regards to Intuition and Scalos, please keep in mind that they IntuitionWM is alpha (still has bugs) and Scalos for Linux is right now even pre-alpha. If you are still interested, let me know whether you want pre-compiled binaries or you want to compile yourself from source code.

With regards to requirements, I use Linux Mint 21 and the videos you might have seen have IntuitionWM and Scalos configured under this distribution. If you use different distribution, I'd still suggest to start with Linux Mint 21 in a VM (VirtualBox, etc) and only then move to other distribution.

**Reply by xgnusr** [reply]
Hello,

Thank you very much for your response, I would prefer precompiled binaries; However, I have started to prepare a virtual machine, following your recommendations to start compiling sources.
I will be very attentive to the evolution of AXruntime; After so many years, I think it is the most reasonable thing to do to have an Amiga running again.
Thank you so much !!!

**Reply by deadwood2** [reply]
Ok, I've made IntuitionWM available here: https://axrt.org/development/Intuition_BETA-1.zip. My recommendation would be to first install it on virtual machine before doing so on your main hardware. For Scalos, I won't be making binaries available for now, but if you are interested in building it from source, here is a discussion I'm having with another person on doing so: https://github.com/deadw00d/AROS/issues/106

Let me know if Intuition worked for you :)

**Reply by xgnusr** [reply]
Works !! It's wonderful to see those windows again.
This is the way !!

Thank you so much !!!

**Comment by markusbkk** [comment]
> Hello,
> 
> Thank you very much for your response, I would prefer precompiled binaries; However, I have started to prepare a virtual machine, following your recommendations to start compiling sources. I will be very attentive to the evolution of AXruntime; After so many years, I think it is the most reasonable thing to do to have an Amiga running again. Thank you so much !!!

Hey there, @xgnusr

Looks like we have the same goal (maybe?)

I'm actually hoping to create a little distro, complete with an installer GUI (naturally also written with Intuition) and a lightweight, transparent Amiga emulator for classic software that cannot be recompiled.

**Reply by deadwood2** [reply]
I can already give you a hint then: there is software called Janus-UAEv1. This is a version of Amiga emulator that is integrated with AROS. If you take it and compile it against AxRuntime (and fix the 64-bit bugs), the you will get very nice integration of Amiga software into Linux through AxRuntime.

**Reply by deadwood2** [reply]
Alternatively there is Janus-UAEv2, which was being worked for 64-bit, has much never code base but is missing the integration bits.

**Reply by xgnusr** [reply]
Hello !!
Thanks for your message.
Actually, my goal is to have an Amiga up and running, taking the right steps.
For me, these steps are:
- Intuition + Scalos as desktop
- compile programs, which were useful to me years ago, with axruntime
- help make the Amiga a desktop computer again.
Of course that desktop would have an emulator (maybe a hardware device...) for 68k games and something else.

TNX markusbkk !

**Comment by markusbkk** [comment]
Curious. Is it "normal" that some MUI attributes aren't defined?

For starters, `MUIA_Slider_Level` didn't exist.

Putting this on top of the source file works, as per code from this [wiki](https://en.wikibooks.org/wiki/Aros/Developer/Zune/CustomClasses).

```c
#ifndef MUIA_Slider_Level
#define MUIA_Slider_Level                   0x8042ae3a /* V4  isg LONG              */
#endif
```

**Reply by deadwood2** [reply]
Looking at mui.h, a few attributes are wrapped in MUI_OBSOLETE #ifdef. I don't know what's the history behind this, but you can get those attributes by defining MUI_OBSOLETE before including mui.h :)

**Comment by markusbkk** [comment]
Messing about with `make contrib-regina` RN.

I got somewhere but now it's complaining that `AROS/contrib/regina/./mt_notmt.c:168: undefined reference to __regina_OS_Unx`

**Reply by deadwood2** [reply]
Here is what comes to my mind: AxRuntime compiles code using real linux compiler, which means certain ported code can think it is being compiled in Linux and not on "AROS" - it's probably due to some #ifdefs in code, which you need to now change.

Note that contrib has not been compiled for AxRuntime (so code adjustements have not been made) except for a few elements.

**Comment by deadwood2** [comment]
BTW, if you recompile any software for AxRuntime and would like to host it "centrally", let me know. I'll host it on axrt.org.

**Reply by markusbkk** [reply]
Absolutely.. 

Currently, I'm experimenting with [D](https://dlang.org/spec/betterc.html). Had to recompile [GDC](https://github.com/D-Programming-GDC/gcc/tree/ci/gcc-10) since the install script only downloads a 2016 release of the GDC compiler.

I still need to figure out how to change the Makefile so it'll place/search **$DEP** inside a subfolder, since it otherwise breaks the build (both D language files and Makefile dependencies use **.d** endings), but manually calling the following works.

```
/usr/local/bin/gdc -fno-druntime -g -O0 -I/usr/include/axrt-4.0 -specs=/usr/lib/x86_64-linux-gnu/axrt-4.0/axrt.specs -fPIC   -c -o hello.o hello.d && /usr/local/bin/gdc -fno-druntime -o hello hello.o /usr/lib/x86_64-linux-gnu/axrt-4.0/axrt.ld -L/usr/lib/x86_64-linux-gnu/axrt-4.0 -specs=/usr/lib/x86_64-linux-gnu/axrt-4.0/axrt.specs -shared -ldl -lm -lpthread
```
My reasoning for using D (instead of C/C++) with AxRuntime is that I think it would be a great boon to both (rather niche) communities. Maybe this way we can have that one "killer app" that popularizes both technologies and D software devs will have a reason to write more D.

EDIT: Ok. First small problem. D doesn't support C header includes. D devs recommend external [tools](https://github.com/atilaneves/dpp) for that. Hooking in a pre-processor with that sort of configuration might also be a hassle. Meh.

**Comment by markusbkk** [comment]
Two days ago I read up on MorphOS having an ObjectiveC API since [3.10](https://www.morphos-team.net/guide/objective-c). 
Since then I've been experimenting with [ObjFW](https://github.com/ObjFW/ObjFW). Unfortunately, I know very little about compiler intrinsics.

```cc -o HelloApp HelloApp.m -I/usr/local/include   -fexceptions -fobjc-exceptions -funwind-tables -fconstant-string-class=OFConstantString -fgnu-runtime  -fexceptions  -Wl,-rpath,/usr/local/lib -L/usr/local/lib -lobjfw -lobjfwrt  -lm -ldl -lpthread /usr/lib/x86_64-linux-gnu/axrt-4.0/axrt.ld -L/usr/lib/x86_64-linux-gnu/axrt-4.0 -specs=/usr/lib/x86_64-linux-gnu/axrt-4.0/axrt.specs -shared -ldl -lm -lpthread -fPIC```

Produces:

>./HelloApp 
<<INFO>>: AxRT 41.6
<<INFO>>: Using absolute paths.
<<INFO>>: AXRT_ROOT environment variable set
<<INFO>>: RUNTIME_ROOT: /home/user/libaxrt/alt-runtimelinux-x86_64-d/bin/runtimelinux-x86_64/AROS/
<<INFO>>: AXRTSYS     : ROOT:home/user/libaxrt/alt-runtimelinux-x86_64-d/bin/runtimelinux-x86_64/AROS/
<<INFO>>: USERSYS     : ROOT:home/user/.axrt/
<<INFO>>: USERHOME    : ROOT:home/user/
<<INFO>>: CURRENT_DIR : ROOT:home/user/obj-c-Example/
<<INFO>>: PROGRAM DIR : ROOT:home/user/obj-c-Example/
<<INFO>>: PROGRAM NAME: HelloApp
[ObjFWRT @ class.m:568] Class HelloApp not found!
Aborted (core dumped)

@deadw00d Any idea why this fails? I assume marrying two runtimes like this won't work out of the box?

**HelloApp.m** source code:

```objc
#import <ObjFW/ObjFW.h>
#import <ObjFW/OFLocale.h>
#import <ObjFW/OFString.h>
#import <sys/utsname.h>

struct utsname buffer;

@interface HelloApp: OFObject <OFApplicationDelegate>
-(void) print;
@end

OF_APPLICATION_DELEGATE(HelloApp)

@implementation HelloApp
- (void)applicationDidFinishLaunching: (OFNotification *)notification
{
        uname(&buffer);

        [OFStdOut writeLine: @"Hello World!"];
        [OFStdOut writeLine: OFApplication.programName];
        [OFStdOut writeLine: OFLocale.languageCode];
        OFString *stringFromUTFString = [[OFString alloc] initWithUTF8String: buffer.version];
        [OFStdOut writeLine: stringFromUTFString];
        [self print];
        [OFApplication terminate];
}

- (void)print
{
    OFLog (@"Hello Print");
}
@end
```

**Reply by deadwood2** [reply]
Hmm, unfortunately I know next to nothing about how ObjectiveC works. I guess there is some loader in the framework part and this can conflict with how AxRuntime applications are build as they use a special loading procedure as well. What you can do is try running it through gdb and see if you get any crash or something.

**Reply by markusbkk** [reply]
I think I figured it out, albeit I don't quite know how to alleviate it yet.
The problem is that AxRuntime obviously uses `main()` whereas the Objective C runtime calls into `NSApplicationMain()`,

**Comment by markusbkk** [comment]
@deadw00d May I have access to the IntuitionWM source?

**Reply by deadwood2** [reply]
Sure, you will find the sources here:

https://github.com/deadw00d/AROS/tree/alt-runtime/arch/all-runtimelinux/intuition

**Reply by markusbkk** [reply]
Awesome, thanks.
I tried the binary release this morning but the README said I'd require parts of MATE and when I installed MATE I just screwed up my Gnome 3 setup under Pop!_OS. Hoping to dissect it myself and use it without any dependencies on DEs.

**Reply by deadwood2** [reply]
Well it doesn't require MATE, but has been tested under MATE and configuration files are also tested under MATE, so if you have different DE, then you indeed possibly need to do work on integration.

**Comment by markusbkk** [comment]
Not really related to the window management inside AxRuntime Intuition but I just found this and thought it might be interesting (if only for its historical value):

https://github.com/RasmusEdgar/dintwm
